packages/network/lib/pki.ts

```diff
@@ -0,0 +1,178 @@
+        this.path = undefined
+      } else if (parsed.pathname.length > 0 &&
+                  !parsed.pathname.endsWith('/') &&
+                  !parsed.pathname.endsWith('*')) {
+        this.path = `${parsed.pathname}/`
+      } else {
+        this.path = parsed.pathname
+      }
+    }
+  }
+
+  path: string | undefined
+  host: string
+  port: number | undefined
+}
+
+export class PkiUrlMatcher {
+  static buildMatcherRule (url: string): ParsedUrl {
+    let parsed = new ParsedUrl(url)
+
+    parsed.host = this.createMatcherValue(parsed.host)
+    parsed.path = this.createMatcherValue(parsed.path)
+
+    return parsed
+  }
+
+  static matchUrl (parsedUrl: ParsedUrl, rule: ParsedUrl | undefined): boolean {
+    if (!parsedUrl || !rule) {
+      return false
+    }
+
+    let ret = new RegExp(rule.host).test(parsedUrl.host)
+
+    if (ret && rule.port) {
+      ret = rule.port === parsedUrl.port
+    }
+
+    if (ret && rule.path) {
+      ret = new RegExp(rule.path).test(parsedUrl.path || '')
+    }
+
+    return ret
+  }
+
+  static createMatcherValue (value: string | undefined): string {
+    if (!value) {
+      return ''
+    }
+
+    let escaped = escapeRegExp(value)
+
+    escaped = escaped.replace(/\\\*/g, '.+') // Replace all instances of '\*' with '.+'
+
+    return `^${escaped}$`
+  }
+}
+
+/**
+ * Defines the certificates that should be used for the specified URL
+ */
+export class UrlPkiCertificates {
+  constructor (url: string) {
+    this.subjects = ''
+    this.url = url
+    this.pathnameLength = new URL(url).pathname.length
+    this.pkiCertificates = new PkiCertificates()
+  }
+  pkiCertificates: PkiCertificates
+  url: string
+  subjects: string
+  pathnameLength: number
+  matchRule: ParsedUrl | undefined
+
+  addSubject (subject: string) {
+    if (!this.subjects) {
+      this.subjects = subject
+    } else {
+      this.subjects = `${this.subjects} - ${subject}`
+    }
+  }
+}
+
+/**
+ * PKI certificates; this is in a data structure that is compatible with the NodeJS TLS API described
+ * at https://nodejs.org/api/tls.html#tls_tls_createsecurecontext_options
+ */
+export class PkiCertificates {
+  ca: Buffer[] = []
+  cert: Buffer[] = []
+  key: PemKey[] = []
+  pfx: PfxCertificate[] = []
+}
+
+export class PemKey {
+  constructor (pem: Buffer, passphrase: string | undefined) {
+    this.pem = pem
+    this.passphrase = passphrase
+  }
+
+  pem: Buffer;
+  passphrase: string | undefined
+}
+
+export class PfxCertificate {
+  constructor (buf: Buffer, passphrase: string | undefined) {
+    this.buf = buf
+    this.passphrase = passphrase
+  }
+
+  buf: Buffer
+  passphrase: string | undefined
+}
+
+export class ClientPkiCertificateStore {
+  private static _instance: ClientPkiCertificateStore
+  private _urlPkicertificates: UrlPkiCertificates[] = []
+
+  public static get Instance () {
+    return this._instance || (this._instance = new this())
+  }
```

Comment:
Refactor usage of this class to not rely on a global singleton instance.

Reply:
Hi @flotwig - we should be able to provide an update here early next week

Reply:
@flotwig Are you happy this is resolved in the latest commit? Can this conversation be resolved?